File: 3955-valid-binary-strings-with-cost-limit/3955-valid-binary-strings-with-cost-limit.go
func generateValidStrings(n int, k int) []string {
	ans := make([]string, 0)

	var dfs func(int, int, bool, string)

	dfs = func(i, c int, p bool, s string) {
		if c > k {
			return
		}

		if i == n {
			ans = append(ans, s)
			return
		}

		dfs(i+1, c, false, s+"0")

		if !p {
			dfs(i+1, c+i, true, s+"1")
		}
	}

	dfs(0, 0, false, "")
	return ans
}

File: 3955-valid-binary-strings-with-cost-limit/3955-valid-binary-strings-with-cost-limit.rs
impl Solution {
    pub fn generate_valid_strings(n: i32, k: i32) -> Vec<String> {
        fn dfs(
            i: i32,
            c: i32,
            p: bool,
            s: String,
            n: i32,
            k: i32,
            ans: &mut Vec<String>,
        ) {
            if c > k {
                return;
            }

            if i == n {
                ans.push(s);
                return;
            }

            dfs(i + 1, c, false, format!("{}0", s), n, k, ans);

            if !p {
                dfs(i + 1, c + i, true, format!("{}1", s), n, k, ans);
            }
        }

        let mut ans = Vec::new();
        dfs(0, 0, false, String::new(), n, k, &mut ans);
        ans
    }
}